File: src/UiPath.CoreIpc/Dtos.cs
using System.Diagnostics.CodeAnalysis;
using System.Text;
using Newtonsoft.Json;
namespace UiPath.CoreIpc;
public class Message
{
    internal Type CallbackContract { get; set; }
    [JsonIgnore]
    public IClient Client { get; set; }
    [JsonIgnore]
    public TimeSpan RequestTimeout { get; set; }
    public TCallbackInterface GetCallback<TCallbackInterface>() where TCallbackInterface : class => 
        Client.GetCallback<TCallbackInterface>(CallbackContract);
    public void ImpersonateClient(Action action) => Client.Impersonate(action);
}
public class Message<TPayload> : Message
{
    public Message(TPayload payload) => Payload = payload;
    public TPayload Payload { get; }
}
record Request(string Endpoint, string Id, string MethodName, string[] Parameters, double TimeoutInSeconds)
{
    internal Stream UploadStream { get; set; }
    public override string ToString() => $"{Endpoint} {MethodName} {Id}.";
    internal TimeSpan GetTimeout(TimeSpan defaultTimeout) => TimeoutInSeconds == 0 ? defaultTimeout : TimeSpan.FromSeconds(TimeoutInSeconds);
}
record CancellationRequest(string RequestId);
record Response(string RequestId, string Data = null, Error Error = null)
{
    internal Stream DownloadStream { get; set; }
    public static Response Fail(Request request, Exception ex) => new(request.Id, Error: ex.ToError());
    public static Response Success(Request request, string data) => new(request.Id, data);
    public static Response Success(Request request, Stream downloadStream) => new(request.Id) { DownloadStream = downloadStream };
    public TResult Deserialize<TResult>(ISerializer serializer)
    {
        if (Error != null)
        {
            throw new RemoteException(Error);
        }
        return (TResult)(DownloadStream ?? serializer.Deserialize(Data ?? "", typeof(TResult)));
    }
}
[Serializable]
public record Error(string Message, string StackTrace, string Type, Error InnerError)
{
    [return: NotNullIfNotNull("exception")]
    public static Error? FromException(Exception? exception)
    => exception is null 
        ? null 
        : new(
            Message: exception.Message, 
            StackTrace: exception.StackTrace ?? exception.GetBaseException().StackTrace, 
            Type: GetExceptionType(exception), 
            InnerError: FromException(exception.InnerException));
    public override string ToString() => new RemoteException(this).ToString();

    private static string GetExceptionType(Exception exception) => (exception as RemoteException)?.Type ?? exception.GetType().FullName;
}
[Serializable]
public class RemoteException : Exception
{
    public RemoteException(Error error) : base(error.Message, error.InnerError == null ? null : new RemoteException(error.InnerError))
    {
        Type = error.Type;
        StackTrace = error.StackTrace;
    }
    public string Type { get; }
    public override string StackTrace { get; }
    public new RemoteException InnerException => (RemoteException)base.InnerException;
    public override string ToString()
    {
        var result = new StringBuilder();
        GatherInnerExceptions(result);
        return result.ToString();
    }
    private void GatherInnerExceptions(StringBuilder result)
    {
        result.Append($"{nameof(RemoteException)} wrapping {Type}: {Message} ");
        if (InnerException == null)
        {
            result.Append("\n");
        }
        else
        {
            result.Append(" ---> ");
            InnerException.GatherInnerExceptions(result);
            result.Append("\n\t--- End of inner exception stack trace ---\n");
        }
        result.Append(StackTrace);
    }
    public bool Is<TException>() where TException : Exception => Type == typeof(TException).FullName;
}
enum MessageType : byte { Request, Response, CancellationRequest, UploadRequest, DownloadResponse }

File: src/Clients/python/UiPath-Ipc-Py/src/uipath_ipc/wire/dtos.py
"""Wire protocol data types matching the .NET CoreIpc wire format."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from enum import IntEnum
from typing import Any


class MessageType(IntEnum):
    Request = 0
    Response = 1
    CancellationRequest = 2
    UploadRequest = 3
    DownloadResponse = 4


@dataclass
class Request:
    Endpoint: str
    Id: str
    MethodName: str
    Parameters: list[str]
    TimeoutInSeconds: float = 0.0

    def to_dict(self) -> dict[str, Any]:
        return {
            "Endpoint": self.Endpoint,
            "Id": self.Id,
            "MethodName": self.MethodName,
            "Parameters": self.Parameters,
            "TimeoutInSeconds": self.TimeoutInSeconds,
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> Request:
        return cls(
            Endpoint=d["Endpoint"],
            Id=d["Id"],
            MethodName=d["MethodName"],
            Parameters=d["Parameters"],
            TimeoutInSeconds=d.get("TimeoutInSeconds", 0.0),
        )

    def __str__(self) -> str:
        return f"{self.Endpoint} {self.MethodName} {self.Id}."


@dataclass
class Error:
    Message: str
    StackTrace: str
    Type: str
    InnerError: Error | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "Message": self.Message,
            "StackTrace": self.StackTrace,
            "Type": self.Type,
            "InnerError": self.InnerError.to_dict() if self.InnerError else None,
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any] | None) -> Error | None:
        if d is None:
            return None
        return cls(
            Message=d["Message"],
            StackTrace=d["StackTrace"],
            Type=d["Type"],
            InnerError=cls.from_dict(d.get("InnerError")),
        )

    @classmethod
    def from_exception(cls, ex: BaseException) -> Error:
        import traceback

        return cls(
            Message=str(ex),
            StackTrace="".join(traceback.format_exception(type(ex), ex, ex.__traceback__)),
            Type=f"{type(ex).__module__}.{type(ex).__qualname__}",
            InnerError=cls.from_exception(ex.__cause__) if ex.__cause__ else None,
        )


@dataclass
class Response:
    RequestId: str
    Data: str | None = None
    Error: Error | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "RequestId": self.RequestId,
            "Data": self.Data,
            "Error": self.Error.to_dict() if self.Error else None,
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> Response:
        return cls(
            RequestId=d["RequestId"],
            Data=d.get("Data"),
            Error=Error.from_dict(d.get("Error")),
        )

    @classmethod
    def fail(cls, request: Request, ex: BaseException) -> Response:
        return cls(RequestId=request.Id, Error=Error.from_exception(ex))

    @classmethod
    def success(cls, request: Request, data: str) -> Response:
        return cls(RequestId=request.Id, Data=data)


@dataclass
class CancellationRequest:
    RequestId: str

    def to_dict(self) -> dict[str, Any]:
        return {"RequestId": self.RequestId}

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> CancellationRequest:
        return cls(RequestId=d["RequestId"])
